### PR TITLE
Add OPE tutorial to vowpalwabbit.org

### PR DESCRIPTION
**Pull request by maxpagels**
We lack a clear tutorial on how to evaluate CB policies offline, before deployment. I suspect many, myself included, make mistakes in this area because VW often doesn't warn you if you are doing the wrong thing.

So I wrote a tutorial, which I am opening as a PR. Note that this needs expert commentary, since I am no expert and some of the text may thus be factually false.

Some questions that I would need a detailed response to, in addition to reviewing the text and letting me know if there are typos, but more importantly, factual errors:

- [x] confirm that doubly robust is the default estimator when using `--eval`
- [x] confirm that the average loss reported by `--eval` is in fact the OPE estimate and makes sense to compare to the average cost calculated from a production policy's generated bandit data
- [x] confirm that only unbiased estimators are the only estimators recommended for doing OPE
- [x] confirm that you can compare `--eval` runs of different unbiased estimators (e.g. IPS vs. DR). Is it OK to look at the losses between these and determine the best policy by picking the one with the lowest average loss?
- [x] confirm that when using `--cb` or `cb_adf` to optimise a policy, the PV loss _and_ holdout loss if `--passes` > 0 is in fact the OPE estimate and makes sense to compare to the average cost calculated from a production policy's generated bandit data, _if and only if_ the estimator is unbiased

If the above assumptions are true, I am wondering if there is any way to train a policy using e.g. DM but have the loss reported be e.g. IPS? I can't come up with a single use case in which you would intentionally want to use a biased estimator _and_ report the average loss using a biased estimator. I can imaging lots of people currently gridsearching (ips,dm,dr), looking at the avg. loss, and choosing the one with the lowest loss, which to my understanding is the wrong thing to do.

### CODE REVIEW

**Comment by pmineiro** [comment]
Max, looks good, thanks for doing this.  Here are some comments:

1. `--eval` is for evaluating policies that are not (necessarily) implemented in vw.  it is driven entirely by the label information in the log.  if you want to test a vw policy you can use `-t`.  many users get confused about this in practice.
2. `--eval` is legacy and only works with `--cb` (not `--cb_adf`).  we don't really use it that much anymore (we evaluate policies using the estimation library https://github.com/VowpalWabbit/estimators )
2. `--eval` doesn't build a policy.  however if you ask for a doubly-robust estimate it will build a reward predictor online.  DR is unbiased and it's conceivable that when the reward is highly predictable this results in a lower variance estimate, so it's a sensible thing to try.  However IPS is simpler, faster, and more easily replicated in other languages.
3. with `--eval` DR is selected by default. 
3. don't load `-i` a policy with `--eval`.  for IPS it will do nothing, but for DR you're actually loading a reward predictor in that case, i'm not sure what it will do, probably it will interpret the policy as a bonkers reward predictor for DR which is technically still unbiased but will suck.
4. Conceivably you can use a biased estimator for OPE because you are typically actually concerned about model selection and a biased estimator might order the policies that you care about more correctly.  However this is too confusing so we don't allow DM.
5. differences of IPS or DR estimates are unbiased estimates of the policy value difference.   the difference of an IPS estimate and a DR estimate is also unbiased (linearity of expectation).
6. the PV and holdout semantics work analogously to supervised learning, in that by evaluating on untrained examples one is estimating generalization performance presuming iid.  however, one can estimate the true expected value of a biased quantity, which is what happens if you used PV or holdout with, e.g., DM.

**Comment by maxpagels** [comment]
Thanks for the comments!

Fixed 4), added text to address 5).

With regards to the rest, I have further questions which I will post here soon. In the meantime, I'm going to try an restructure the text since you get the feeling that `--eval` is the only recommended approach if using the CLI, when in fact is kind of a legacy feature.

**Comment by maxpagels** [comment]
Right, incoming wall of text :) I've probably asked these same questions before, but I know lots of people are interested in the answer so I'll asking again to make absolutely sure I get the tutorial right.

Lets say I'm a data scientist that has access to bandit data produced by a logging policy currently in production. There are two scenarios, either I plan on deploying a new model into production where it learns in an online fashion/incrementally ("online scenario"), or I train a model in batch fashion, deploy that and the deployed model doesn't change until I do a new deployment ("batch scenario").

Now, in both these settings, I am interested in knowing what combo of `cb_type`, regularisation, `lr`, etc hyperparameters would result in something that is estimated to work better than what I currently have in prod.

Are the following approaches valid:

- online scenario: I do hyperparameter search, i.e. running `vw --cb <n_arms> --cb_type <cb_type> <other options> -d <logging_policy_dataset>` for each hyperparam combo, and record the PV loss for each. I then compare the PV losses against the realised average loss in the logged dataset. if one of the recorded losses is lower than the realised loss in the logged data, I have found a policy that is estimated to work better than the one currently in prod. Note that my gridsearch might include `--cb_type ips` in one combination and `--cb_type dr` in another. Is that still OK? I'm ruling out the use of DM because of what you mentioned, you might technically want to use it but it's a bit confusing.

- batch scenario: I gridsearch as before, but train on several passes and keep record of the holdout losses. I compare these to the realised average loss in the logged dataset.  if one of the recorded losses is lower than the realised loss in the logged data, I have found a policy that is estimated to work better than the one currently in prod. Again, I'm excluding DM here.

- batch scenario: I split the logged bandit data such that I get a sufficiently large training set, and a holdout set of examples that, temporally speaking, arrived later than the stuff in the training set.  I gridsearch as before, but save each model to disk and then use `-t` against the separate holdout data file coming from the logging policy. I check the reported losses (does `-t` report PV loss here, too, of straight up average loss?). If the average loss from any one of my testing runs is lower than the average loss as calculated directly from the holdout file, I have found a policy that works better than that in prod. Again, DM is excluded here.

**Comment by maxpagels** [comment]
I rewrote to include scenarios and downplay the `--eval` option. I would need confirmation on the following before making a "final" edit and then the whole thing can be read through again.

If I've trained two policies offline, one using IPS and the other DR, and I test against a test set and IPS gets 0.5 and DR 0.3, is it fair to compare these to values and say DR is "better" in this case?

**Comment by maxpagels** [comment]
@pmineiro ready for review again. And do you have a comment on the previous question?

**Comment by pmineiro** [comment]
> @pmineiro ready for review again. And do you have a comment on the previous question?
 
Which one? :)

> If I've trained two policies offline, one using IPS and the other DR, and I test against a test set and IPS gets 0.5 and DR 0.3, is it fair to compare these to values and say DR is "better" in this case?

tl;dr: "fair", but not advised.

How you train a policy, in terms of evaluation comparison, doesn't matter.  Maybe you pulled a weight vector out of your posterior.  However when comparing two policies to evaluate (estimate) the policy value difference, it's most sane to compare them using the same methodology.  In theory the bias of both IPS and DR is zero and therefore the bias of the difference where either strategy is used to evaluate each component of the difference is still zero.  In practice I would personally trust more a comparison where both are evaluated the same way (e.g., both IPS).   Even better would be to form a confidence interval around the policy value difference itself, which can benefit from pairwise comparison.  A simple way to do this would be to form the IPS policy value difference for each datapoint ( ( w_policy1 - w_policy2 ) * r ) and then compute bootstrap mean distribution (see https://pypi.org/project/bootstrapped/ ).  For a more advanced way to directly construct a confidence sequence on the policy value difference see section 5.3 of  https://arxiv.org/abs/2102.09540v1 .

**Comment by maxpagels** [comment]
Hi @pmineiro,

That makes full sense to me, updated the caveat that mixing loss metrics isn't advised.

I also understand the difference between train/eval, it's just that VW specifically doesn't allow you to use the CLI to get e.g. the IPS loss of a DR-trained model. In other words, if you do something like `vw --cb 10 --cb_type ips -f model.vw -d mnist.txt.gz` followed by ` vw -i model.vw -t -d mnist-holdout.txt --cb_type dr` you will get an error. That's why my question was framed as one of training, which it technically isn't. VW doesn't allow mixing and matching IPS/DR/DM for training and loss reporting, at least not in the CLI.

Either way, I'm pretty much done now and I think the text is good to go for another read.

**Comment by maxpagels** [comment]
Random side note: once the estimators Python lib is in the state that it is relatively mature, I can make another tutorial for that.

**Comment by maxpagels** [comment]
Is there anything to tweak still on this, or is the content sensible?

**Comment by jackgerrits** [comment]
Thanks so much for producing this @maxpagels! Awesome stuff honestly. I'll get this deployed to the live site tomorrow.

**Comment by jackgerrits** [comment]
This has now been merged
